geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/index/QueryPlanner.scala

```diff
@@ -74,28 +73,40 @@ case class QueryPlanner(sft: SimpleFeatureType,
    */
   def runQuery(query: Query, strategy: Option[Strategy] = None): SFIter = {
     val (plans, numClauses) = getStrategyPlans(query, strategy, log)
-    val dedupe = numClauses > 1 || plans.exists(_.plan.hasDuplicates)
+    // don't deduplicate density queries, as they don't have dupes but re-use feature ids in the results
+    val dedupe = !query.getHints.isDensityQuery && (numClauses > 1 || plans.exists(_.plan.hasDuplicates))
     executePlans(query, plans, dedupe)
   }
 
   /**
-   * Execute a query plan. Split out to allow for easier testing.
+   * Execute a sequence of query plans
    */
   private def executePlans(query: Query, strategyPlans: Seq[StrategyPlan], deduplicate: Boolean): SFIter = {
-    val features = strategyPlans.iterator.ciFlatMap { sp =>
+    def scan(sps: Seq[StrategyPlan]): SFIter = sps.iterator.ciFlatMap { sp =>
       sp.strategy.execute(sp.plan, acc, log).map(sp.plan.kvsToFeatures)
     }
 
-    val dedupedFeatures = if (deduplicate) new DeDuplicatingIterator(features) else features
+    def dedupe(iter: SFIter): SFIter = if (deduplicate) new DeDuplicatingIterator(iter) else iter
 
-    //noinspection EmptyCheck
-    val sortedFeatures = if (query.getSortBy != null && query.getSortBy.length > 0) {
-      new LazySortedIterator(dedupedFeatures, query.getHints.getReturnSft, query.getSortBy)
+    // noinspection EmptyCheck
+    def sort(iter: SFIter): SFIter = if (query.getSortBy != null && query.getSortBy.length > 0) {
+      // sort will self-close itself
+      new LazySortedIterator(iter, query.getHints.getReturnSft, query.getSortBy)
     } else {
-      dedupedFeatures
+      // wrap in a self-closing iterator to mitigate clients not calling close
+      SelfClosingIterator(iter)
     }
 
-    adaptIterator(query, sortedFeatures)
+    def expand(iter: SFIter): SFIter = if (query.getHints.isTemporalDensityQuery) {
+      val encode = query.getHints.containsKey(RETURN_ENCODED)
+      TemporalDensityIterator.reduceTemporalFeatures(iter, query.getHints.getReturnSft, encode)
```

Comment:
For consistency, consider refactoring the TDI object's reduceFeatures method to just get the query.  

Potentially, this could head to a much more pluggable framework...  The object could speak for whether or not it can override behavior and the reduceFeatures function would fit an interface.